### DISCUSSION TITLE
Does `needs_recompile` work in build isolation?

### DISCUSSION THREAD

**Discussion by ketozhang** [pybind/pybind11#4345] [Q&A]
[Docs](https://pybind11.readthedocs.io/en/stable/compiling.html#building-with-setuptools) does not mention this, but the source code does.

https://github.com/pybind/pybind11/blob/296615ad34f9d8f680efbab22553881ad9843063/pybind11/setup_helpers.py#L391-L394

In this minimal `setup.py` file (`pyproject.toml` not shown),

```python
ParallelCompile(default=0, needs_recompile=naive_recompile).install()

ext_modules = Pybind11Extension(
    "myextpkg",
    sources=["src/myextpkg.cpp"]
)

setup(ext_modules=ext_modules)
```

After building once so the object files exists in `src/myextpkg.cpython-***.so`, I recompile with...

```
$ time pip install -e .
23.71s user 3.72s system 99% cpu 27.696 total
```

```
$ time pip install --no-build-isolation -e .
3.22s user 1.06s system 116% cpu 3.679 total
```

**Comment by LukeLIN-web** [comment]
So you mean it can make it work quicker?
pybind11 doc emits many details. I often feel confused when I read it.
